types/scene/surface_tree.c

```diff
@@ -0,0 +1,206 @@
+	wl_list_remove(&surface_tree->surface_destroy.link);
+	wl_list_remove(&surface_tree->surface_commit.link);
+	wl_list_remove(&surface_tree->surface_new_subsurface.link);
+	free(surface_tree);
+}
+
+static void surface_tree_handle_tree_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_scene_surface_tree *surface_tree =
+		wl_container_of(listener, surface_tree, tree_destroy);
+	surface_tree_destroy(surface_tree);
+}
+
+static void surface_tree_handle_surface_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_scene_surface_tree *surface_tree =
+		wl_container_of(listener, surface_tree, surface_destroy);
+	wlr_scene_node_destroy(&surface_tree->tree->node);
+}
+
+static const struct wlr_addon_interface surface_tree_addon;
+
+static struct wlr_scene_surface_tree *surface_tree_from_subsurface(
+		struct wlr_scene_surface_tree *parent,
+		struct wlr_subsurface *subsurface) {
+	struct wlr_addon *addon = wlr_addon_find(&subsurface->surface->addons,
+		parent, &surface_tree_addon);
+	assert(addon != NULL);
+	struct wlr_scene_surface_tree *surface_tree =
+		wl_container_of(addon, surface_tree, surface_addon);
+	return surface_tree;
+}
+
+static void surface_tree_reconfigure(
+		struct wlr_scene_surface_tree *surface_tree) {
+	struct wlr_surface *surface = surface_tree->surface;
+
+	struct wlr_scene_node *prev = NULL;
+	struct wlr_subsurface *subsurface;
+	wl_list_for_each(subsurface, &surface->current.subsurfaces_below,
+			current.link) {
+		struct wlr_scene_surface_tree *child =
+			surface_tree_from_subsurface(surface_tree, subsurface);
+		if (prev != NULL) {
+			wlr_scene_node_place_above(&child->tree->node, prev);
+		}
+		prev = &child->tree->node;
+
+		wlr_scene_node_set_position(&child->tree->node,
+			subsurface->current.x, subsurface->current.y);
+	}
+
+	if (prev != NULL) {
+		wlr_scene_node_place_above(&surface_tree->scene_surface->node, prev);
+	}
+	prev = &surface_tree->scene_surface->node;
+
+	wl_list_for_each(subsurface, &surface->current.subsurfaces_above,
+			current.link) {
+		struct wlr_scene_surface_tree *child =
+			surface_tree_from_subsurface(surface_tree, subsurface);
+		wlr_scene_node_place_above(&child->tree->node, prev);
+		prev = &child->tree->node;
+
+		wlr_scene_node_set_position(&child->tree->node,
+			subsurface->current.x, subsurface->current.y);
+	}
+}
+
+static void surface_tree_handle_surface_commit(struct wl_listener *listener,
+		void *data) {
+	struct wlr_scene_surface_tree *surface_tree =
+		wl_container_of(listener, surface_tree, surface_commit);
+
+	// TODO: only do this on subsurface order or position change
+	surface_tree_reconfigure(surface_tree);
+}
+
+static void surface_tree_addon_destroy(struct wlr_addon *addon) {
+	struct wlr_scene_surface_tree *surface_tree =
+		wl_container_of(addon, surface_tree, surface_addon);
+	wlr_scene_node_destroy(&surface_tree->tree->node);
+}
+
+static const struct wlr_addon_interface surface_tree_addon = {
+	.name = "wlr_scene_surface_tree",
```

Comment:
Not really related to this PR, but why does this `name` field exist? Just to make poking around at `wlr_addon_set` with a debugger easier?

Reply:
Yeah, it's mainly for making it easier to figure out what an addon represents when debugging (addons are kind of opaque).